### PR TITLE
Differential privacy

### PR DESCRIPTION
**Pull request by shengy90**
1. Implement Differential Privacy with Faraday
2. Added notebook tutorial on differential privacy

### CODE REVIEW

File: src/models/faraday/model.py [hunk 1]
```diff
@@ -132,6 +138,30 @@ def __init__(
         self.lower_quantile = lower_quantile
         self.upper_quantile = upper_quantile
 
+        if differential_privacy:
+
+            if max_grad_norm is None:
```
Comment (anguschadney) [inline src/models/faraday/model.py:143]: in terms of checking params, can `epsilon` be any value?  I thought it needed to be positive for some reason...

File: src/models/faraday/model.py [hunk 2]
```diff
@@ -132,6 +138,30 @@ def __init__(
...
+                raise ValueError("Max grad norm must be set for DP training")
+            elif max_grad_norm > 1 or max_grad_norm < 0:
+                raise ValueError("Max grad norm must be between 0 and 1")
+
+            logger.info("🔒 Differential Privacy Enabled")
+            logger.info("🔒 Epsilon: {epsilon}, Delta: {delta}")
+            logger.info(
+                "🔒 Note: to satisfy definition of"
+                "differential privacy, Delta: {delta}"
+                "must be 1 < N where N is the size of"
```
Comment (anguschadney) [inline src/models/faraday/model.py:153]: is this correct? the default value of `delta` is 1e-5 above...

File: src/models/faraday/model.py [hunk 3]
```diff
@@ -132,6 +138,30 @@ def __init__(
...
+                "the training dataset"
+            )
+            self.differential_privacy = differential_privacy
+            self.max_grad_norm = max_grad_norm
+            self.epsilon = epsilon
+            self.delta = delta
+
+            self.privacy_engine = PrivacyEngine(secure_mode=False)
+            # Check that everything is valid for DP training
+            assert ModuleValidator.validate(self, strict=True) == []
```
Comment (anguschadney) [inline src/models/faraday/model.py:163]: ah I assume this does the validation?

**Review by anguschadney — APPROVED** [review]
looks good, minor comments to improve my understanding!

File: src/models/faraday/model.py (hunk 1 above)
Comment (shengy90) [reply src/models/faraday/model.py:143]: Yeah it has to be, but with invalid value Opacus will throw up an error on its own that I don't think we need to check. Opacus allows valid positive float values of `max_grad_norm` but we should stick it to max 1.0 as good privacy practice.

File: src/models/faraday/model.py (hunk 3 above)
Comment (shengy90) [reply src/models/faraday/model.py:163]: Yes - that's to check that the model class is set up properly for DP. Opacus has a handy tool to check that all required elements are there for DP to run properly.

**Pull request merged**